Issue 227 Remove stateful elements from server-side datatable communication

Laukkala opened the pull request:
## Description
Simplifies communication between AJS_01 and ZEP_01 regarding server-side datatables.
Previous implementation used angular object registries to both receive and send data to UI when using DPL datatables. The UI and Server were both listening to changes in stateful elements to trigger some actions.
These elements have been removed from the server-side, and now data is provided in a uniform way.

This version uses a different method to send data to UI, and has formatted some values to be the same as is expected by the UI's datatable implementation. 

### Changes
Removed AJAXResponse & Watcher
Removed DTAnchorReady & Watcher
Removed DTServerupdate & Watcher
Removed PageRefresh & Watcher
AJAXRequest is still in use.

Table data is sent to UI via PARAGRAPH_UPDATE_OUTPUT messages instead of ANGULAR_OBJECT_UPDATE AJAXResponse messages.
Pagination requests are received from UI via PARAGRAPH_UPDATE_RESULT messages instead of ANGULAR_OBJECT_UPDATED AJAXRequest messages. Internally still uses AngularObjectRegistry to access the correct instance of DTTableDatasetNG.
AJAXRequest angular objects are now additionally identified with their associated paragraphId. 

Adds a header object to the paginated table data for use on UI side.
Adds a ResultsTotal and ResultsFiltered fields to table data for use on UI side.
Adds a draw value to table data for use on UI side.

Added new JSONTABLE InterpreterResult type.



## Checklists

<!-- Fill check boxes before submitting the pull request. -->

### Testing

#### General

- [x] I have checked that my test files and functions have meaningful names.
- [x] I have checked that each test tests only a single behavior.
- [x] I have done happy tests.
- [x] I have tested only my own code.
- [x] I have tested at least all public methods.

#### Assertions

- [x] I have checked that my tests use assertions and not runtime overhead.
- [x] I have checked that my tests end in assertions.
- [x] I have checked that there is no comparison statements in assertions.
- [x] I have checked that assertions are in tests and not in helper functions.
- [x] I have checked that assertions for iterables are outside of for loops and both sides of the iteration blocks.
- [x] I have checked that assertions are not tested inside consumers.

#### Testing Data

- [x] I have tested algorithms and anything else with the possibility of unbound growth.
- [x] I have checked that all testing data is local and fully replaceable or reproducible or both.
- [x] I have checked that all test files are standalone.
- [x] I have checked that all test-specific fake objects and classes are in the test directory.
- [x] I have checked that my tests do not contain anything related to customers, infrastructure or users.
- [x] I have checked that my tests do not contain non-generic information.
- [x] I have checked that my tests do not do external requests and are not privately or publicly routable.

#### Statements

- [x] I have checked that my tests do not use throws for exceptions.
- [x] I have checked that my tests do not use try-catch statements.
- [x] I have checked that my tests do not use if-else statements.

#### Java

- [x] I have checked that my tests for Java uses JUnit library.
- [x] I have checked that my tests for Java uses JUnit utilities for parameters.

#### Other

- [x] I have only tested public behavior and not private implementation details.
- [x] I have checked that my tests are not (partially) commented out.
- [x] I have checked that hand-crafted variables in assertions are used accordingly.
- [x] I have tested [Object Equality](https://docs.oracle.com/javase/6/docs/api/java/lang/Object.html#equals%28java.lang.Object%29).
- [x] I have checked that I do not have any manual tests or I have a valid reason for them and I have explained it in the PR description.

### Code Quality

- [x] I have checked that my code follows metrics set in Procedure: Class Metrics.
- [x] I have checked that my code follows metrics set in Procedure: Method Metrics.
- [x] I have checked that my code follows metrics set in Procedure: Object Quality.
- [x] I have checked that my code does not have any NULL values.
- [x] I have checked my code does not contain FIXME or TODO comments.

Laukkala committed:
Removed stateful elements, added headers to AJAXResponse

Laukkala committed:
DTHeader can now turn itself to JSON, and AJAXResponse uses the JSON version of the headers.

Laukkala committed:
AJAXResponse now contains the headers as JSON.

Laukkala committed:
Appended AJAXRequest and AJAXResponse with paragraphIds in DTTableDatasetNg

Removed PARAGRAPH_OUTPUT_UPDATED message being sent from DTTableDatasetNg

Now Server sends ANGULAR_OBJECT_UPDATE AJAXResponse as soon as it receives output from DPL

Laukkala committed:
Changed updatePage() method to take parameters instead of an AJAXRequest object.
Now handleAjaxRequest() applies defaults

Laukkala committed:
DPL data updates now come via PARAGRAPH_UPDATE_OUTPUT instead of ANGULAR_OBJECT_UPDATE messages

Laukkala committed:
Removed AJAXResponse angular object entirely

Laukkala committed:
Combined updateData and updatePage to the same method

Laukkala committed:
Added a new InterpreterResult type, JSONTABLE. DPL results are now provided using this output type.

Laukkala committed:
Added new client-to-server websocket event  (PARAGRAPH_UPDATE_RESULT) for requesting data with different pagination settings. This replaces AJAXRequests that were previously sent via ANGULAR_OBJECT_UPDATED websocket messages.

Laukkala committed:
Merge branch 'teragrep:main' into issue-227-remove-stateful-elements

Laukkala committed:
Renamed "datalength" to "recordsTotal" in DTTableDatasetNG DTNetResponse()

Added recordsFiltered to DTTableDatasetNG DTNetResponse() to notify UI about how many rows the transmitted DPL data has.

Laukkala committed:
Added proper parsing for "start" and "length" fields for Doubles on Server side

Laukkala committed:
Removed unnecessary ping message on updateParagraphResult

Laukkala committed:
Update paragraph result now parses search value properly

Laukkala committed:
UpdateParagraphResult now expects a Draw value. UPDATE_PARAGRAPH_OUTPUT also now responds with the draw value.

Laukkala committed:
InterpreterGroupId is now generated automatically based on user name and given note ID.

Added some comments

Laukkala committed:
Made RecordsTotal and RecordsFiltered match with what is expected by Datatables.js:
- RecordsTotal is the total number of rows in the dataset.
- RecordsFiltered is the number of matching rows shown when a search string is provided
RecordsFiltered now does not update on pagination, but only when a search string is provided, as is expected by Datatables.js

Laukkala committed:
Removed unused watchers

eemhu commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTHeader.java:
Improvement idea: Add ctor for `schema` and change static methods to non-static `xml()` and `json()`

Laukkala replied:
Implemented in commit [1689d0f](https://github.com/teragrep/zep_01/pull/230/commits/1689d0ff747daf6d47cb0c2fb0fb2d3b0ebbdf92)

Refactoring these methods and thei usages within ZEP_01 to be non-static didn't cause any issues within ZEP_01, but if there is any other code in any other component that references these static methods, this might be a breaking change.

eemhu commented on spark/dpl-interpreter/src/test/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTHeaderTest.java:
```diff
@@ -76,5 +76,10 @@ public void testDTHeader() throws ParserConfigurationException, TransformerExcep
     }
+    @Test
+    public void testJSONDTHeader() throws ParserConfigurationException, TransformerException {
+        String e = "{\"_time\":\"\",\"_raw\":\"\",\"index\":\"\",\"sourcetype\":\"\",\"host\":\"\",\"source\":\"\",\"partition\":\"\",\"offset\":\"\"}";
```
should build the expected json object instead of using string

Laukkala replied:
Fixed in commit [0658a37](https://github.com/teragrep/zep_01/pull/230/commits/0658a37eec5eac9e0057af8554f59a2044d37eeb)

eemhu commented on spark/dpl-interpreter/src/test/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNgTest.java:
```diff
@@ -202,22 +210,50 @@ public void testAJAXResponse() {
+        JsonObject headers = Json.createReader(new StringReader(DTHeader.schemaToJsonHeader(testSchema))).readObject();;
+            JsonObject response = DTTableDatasetNg.DTNetResponse(formated, headers, 1, datasetAsJSON.size(),formated.size());
+
+            assertEquals("" +
```
same here, these should be built with the JsonObjectBuilder instead of strings

Laukkala replied:
Fixed in commit [0658a37](https://github.com/teragrep/zep_01/pull/230/commits/0658a37eec5eac9e0057af8554f59a2044d37eeb)

eemhu commented on zeppelin-server/src/main/java/com/teragrep/zep_01/service/NotebookService.java:
```diff
@@ -662,6 +662,11 @@ public void updateParagraph(String noteId,
 
+  public void updateParagraphResult(String noteId, String paragraphId, String interpreterGroupId,int draw, int start, int length, String searchString,ServiceContext context, ServiceCallback<AngularObject> callback) throws IOException {
+    // Make an AJAXRequest based on given information
+    updateAngularObject(noteId,paragraphId,interpreterGroupId,"AJAXRequest_"+paragraphId,"{\"draw\":"+draw+",\"start\":"+start+",\"length\":"+length+",\"search\":{\"value\":\""+searchString+"\",\"regex\":false}}",context,callback);
```
also here with the string concat

Laukkala replied:
This required adding a dependency to jakarta.json to zeppelin-server, as ZEP_01 has many different modules each with their own dependencies, but I imagine that it shouldn't be a problem. 
Fixed in commit [e94b71d](https://github.com/teragrep/zep_01/pull/230/commits/e94b71dd43609a0513d658d130b0d908c70c9607)

Laukkala committed:
Changed JSON string concatenation to use Jakarta.Json instead in pth_07 tests.

Laukkala committed:
Added jakarta.JSON as a dependency to zeppelin-server module's pom.xml and replaced JSON string concatenation with use of jakarta.JSON

Laukkala committed:
Added a constructor with a schema to DTHeader, Renamed DTHeader's methods and made them non-static.

eemhu commented on spark/dpl-interpreter/src/test/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNgTest.java:
```diff
@@ -212,25 +211,45 @@ public void testAJAXResponse() {
+                    .build();
+            dataBuilder.add(rowJson);
+        }
+        JsonArray data = dataBuilder.build();
```
assert that dataBuilder has correct amount of rows

Laukkala replied:
Fixed in commit [3b5e49e](https://github.com/teragrep/zep_01/pull/230/commits/3b5e49e07fce4d8ece839fc1886bd60e61975038)

Laukkala committed:
Added an assertion for correct number of rows after iteration in DTTableDatasetNgTest

eemhu commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -72,19 +73,12 @@ public final class DTTableDatasetNg extends AbstractUserInterfaceElement {
-    private final AngularObject<Boolean> pageRefreshAngularObject;
-    private final AngularObject<Integer> DTServerUpdateAngularObject;
 
     private List<String> datasetAsJSON = null;
```
Should make an issue about making these objects immutable.

Laukkala replied:
Created a new issue #234 for refactoring

eemhu commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -174,8 +118,12 @@ void refreshPage() {
-            updatePage(ajaxRequest);
+            // Apply defaults if parameters are not provided
+            int start = (ajaxRequest.getStart() == null) ? 0 : ajaxRequest.getStart();
+            int length = (ajaxRequest.getLength() == null) ? 25 : ajaxRequest.getLength();
```
Same here, new issue about removing nulls from AjaxRequest.

Laukkala replied:
Created issue https://github.com/teragrep/zep_01/issues/235 for refactoring

eemhu commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -249,40 +173,54 @@ public void setParagraphDataset(Dataset<Row> rowDataset) {
+            getInterpreterContext().out().write(outputContent);
+            getInterpreterContext().out().flush();
+        } catch (java.io.IOException e) {
+            LOGGER.error(e.toString());
```
Could these errors be handled? Perhaps a new issue also?

Laukkala replied:
Looks like the reasons this catch exists at all is due to:
1. InterpreterOutput.write() can throw an IOException when it has an InterpreterOutputChangeListener object assigned to it (But currently that object is never assigned)
2. InterpreterOutput.flush() inherits the IOException from ByteArrayOutputStream.flush(). (But ByteArrayOutputStream.flush() can never throw an IOException since it's flush() method's body is empty)

Looks to me we could simply remove the IOExceptions from InterpreterOutput.write() and InterpreterOutput.flush(), so we don't have to catch them here.

It's out of scope for this PR, but I made a new issue for this refactoring: https://github.com/teragrep/zep_01/issues/236

eemhu commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -249,40 +173,54 @@ public void setParagraphDataset(Dataset<Row> rowDataset) {
         // TODO these all decode the JSON, it refactor therefore to decode only once
         // searching
-        List<String> searchedList = DTSearch.search(datasetAsJSON, currentSearchString);
+        List<String> searchedList = DTSearch.search(datasetAsJSON, searchString);
```
Other DT-prefixed objects should also be refactored to not include static methods

Laukkala replied:
Fixed in commit [97de6b1](https://github.com/teragrep/zep_01/pull/230/commits/97de6b1185ebf2d5e53f8de7f3f86ea83847f625)

eemhu replied:
should the `rowLists` be ctor arguments instead?

Laukkala replied:
Added the rowLists to the constructor in this commit: [a100cab](https://github.com/teragrep/zep_01/pull/230/commits/a100cabf296d4b66d165635a3d2160e462b3366f)

I had to create new instances of the DT-prefixed objects during the call to searchAndPaginate(), because they are operating on a mutable variable "datasetAsJSON" (which is set in the method setParagraphDataset().
Refactoring the setParagraphDataset would be a pretty large undertaking, so it's probably not in scope for this issue.

eemhu commented on spark/dpl-interpreter/src/test/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNgTest.java:
```diff
@@ -202,24 +205,76 @@ public void testAJAXResponse() {
+        String paragraphId = "testParag";
+        String noteId ="testNoteId";
+        String angularObjectName = "AJAXRequest_"+paragraphId;
+        String angularObjectContent = "{\"start\":0,\"length\":25,\"search\":{\"value\":\"\",\"regex\":\"false\"}}";
```
should this also be generated with jakarta?

Laukkala replied:
Fixed in commit [0c30ab6](https://github.com/teragrep/zep_01/pull/230/commits/0c30ab6c394e0e5f98e18a93776981fbb7dad2db)

eemhu commented on zeppelin-server/src/main/java/com/teragrep/zep_01/socket/NotebookServer.java:
```diff
@@ -1096,18 +1099,49 @@ public void onSuccess(Note newNote, ServiceContext context) throws IOException {
+  private void updateParagraphResult(NotebookSocket conn,
+                                     ServiceContext context,
+                                     Message fromMessage) throws IOException {
+    final String noteId = (String) fromMessage.get("noteId");
```
Can `get` return null?

Laukkala replied:
It can, in case frontend sends a malformed request that doesn't contain a "noteId" field. Same goes for "paragraphId".

Zeppelin is supposed to be able to handle nulls for noteIds and paragraphIds, but I added a null check with a specific exception message for these strings so that UI gets some feedback on their malformed requests in commit [b197d8f](https://github.com/teragrep/zep_01/pull/230/commits/b197d8f0c752380c0f1d187df798b829fab5bbe6)

Laukkala committed:
Refactored DTSearch and DTOrder to be non-static classes.

Laukkala committed:
Using Jakarta instead of a JSON string in DTTableDatasetNgTest.AjaxRequestToJsonTest.

Laukkala committed:
Added a null check to updateParagraphResult for strings in case frontend sends bad parameters.

Laukkala committed:
Added rowList to constructor on DTOrder, DTPagination and DTSearch

Laukkala committed:
Added a decorator for ValidatedMessage, that compares the JSON structure of the underlying message to a user-defined JSON structure.
The message is only valid if the message contains every key in the same hierarchical position as in the user defined JSON structure.

Laukkala commented:
Added a ValidatedMessage decorator in commit [1ab1adc](https://github.com/teragrep/zep_01/pull/230/commits/1ab1adc5a9cd07add6608e60271af8a5e6e8a792)

The decorator ended up being quite complicated, since it needs to be able to validate the positions of the keys within the hierarchy.

In practice, the validation is comparing two JsonObjects with arbitrary Json contents, and returns true if one object contains every key that is present in the other object, in the same position.

If there is some premade method in the library such along the lines of JsonObject.containsAllKeys(JsonObject other), that would have been great

eemhu commented on zeppelin-common/src/main/java/com/teragrep/zep_01/common/ValidatedMessage.java:
```diff
@@ -0,0 +1,174 @@
+    for (String requiredKey : toCompare.keySet()) {
+      JsonValue valueToValidate = toValidate.get(requiredKey);
+      JsonValue valueToCompare = toCompare.get(requiredKey);
+      if(valueToValidate == null){
```
you can use containsKey instead of null check

eemhu commented:
The validatedMessage object does seem a bit complicated, was the plan to use it elsewhere as well?

Laukkala committed:
Replaced null check with JsonObject.containsKey()

Laukkala committed:
Changed ValidatedMessage to only check for values required by UPDATE_PARAGRAPH_RESULT

Laukkala commented:
> The validatedMessage object does seem a bit complicated, was the plan to use it elsewhere as well?

Refactored ValidatedMessage with @eemhu, as it was only planned to use it in this one instance, so it need not be generic.

eemhu reviewed (commented):
Add `equals()` and `hashCode()` implementations and tests for the new objects, otherwise ok!

Laukkala committed:
Added equals() and hashcode() overrides to new objects and equalsVerifier's testContract() to test files.

Laukkala commented:
> Add `equals()` and `hashCode()` implementations and tests for the new objects, otherwise ok!

Added equals and hashCode to new objects along with EqualsVerifier in commit [7d27a8e](https://github.com/teragrep/zep_01/pull/230/commits/7d27a8ef03af2f6b83c25410b6ad095000071575)

eemhu reviewed (dismissed):
LGTM

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTHeader.java:
```diff
@@ -100,4 +108,29 @@ public static String schemaToHeader(StructType schema) throws TransformerExcepti
+        Iterator<StructField> it = schema.iterator();
+        while(it.hasNext()) {
+            StructField column = it.next();
+            builder.add(column.name(),"");
```
builder uses empty values for keys. is there a reason not to use JsonArrayBuilder?

Laukkala replied:
Changed this to JsonArray, as we only need keys.
[e0c9d80](https://github.com/teragrep/zep_01/pull/230/commits/e0c9d80eb76e707df8fc6024c856001093c1356a)

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -45,6 +45,7 @@
  */
 package com.teragrep.pth_07.ui.elements.table_dynamic;
 
+import com.google.gson.Gson;
```
why Gson here and jakarta-json aka javax.json in DTHeader? I would use onky javax.json here

Laukkala replied:
Removed references to Gson from DTTableDatasetNg as it had no usages. Also removed JsonSyntaxException as it looks like a leftover from legacy code in commit [67f57e0](https://github.com/teragrep/zep_01/pull/230/commits/67f57e04bccd6b742ac06f11de3b0941083d3738)

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -174,8 +117,12 @@ void refreshPage() {
     void handeAJAXRequest(AJAXRequest ajaxRequest) {
```
nulls are checked already in AJAXRequestWatcher.

please refactor AJAXRequestWatcher to use jakarta-json and to produce null free object or stub if needed

Laukkala replied:
AjaxRequestWatcher now uses jakarta-json in commit [04ef0f8](https://github.com/teragrep/zep_01/pull/230/commits/04ef0f88427f215f902c4b6c67003a478589bf2b)

Replaced AjaxRequest object with a JsonObject instead, as it looks like the AjaxRequest class is simply a data holder that needed to exist to allow for Gson to deserialize JSON for validation purposes.

Along with other changes in this PR, AjaxRequest no longer has any usages outside tests, so it could be removed.

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -174,8 +117,12 @@ void refreshPage() {
+            String searchString = (ajaxRequest.getSearch().getValue() == null) ? "" : ajaxRequest.getSearch().getValue();
+            updatePage(start,length,searchString, draw);
         }
         catch (JsonSyntaxException e) {
```
JsonSyntaxException is thrown in AJAXRequestWatcher not here why catching it here?

Laukkala replied:
Removed JsonSyntaxException as it looks like a leftover from legacy code in commit [67f57e0](https://github.com/teragrep/zep_01/pull/230/commits/67f57e04bccd6b742ac06f11de3b0941083d3738)

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -96,74 +89,24 @@ public final class DTTableDatasetNg extends AbstractUserInterfaceElement {
-        );
-
-        DTServerUpdateAngularObject.addWatcher(new DTServerUpdateWatcher(interpreterContext));
+        this.gson = new Gson();
```
gson is not used here, why new Gson()?

Laukkala replied:
Leftover from a previous implementation, removed in commit [67f57e0](https://github.com/teragrep/zep_01/pull/230/commits/67f57e04bccd6b742ac06f11de3b0941083d3738) as it has no usages.

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -304,12 +243,15 @@ static JsonArray dataStreamParser(List<String> data){
+    static JsonObject DTNetResponse(JsonArray data, JsonObject datasetAsJSONSchema, int draw, int recordsTotal, int recordsFiltered){
         try{
             JsonObjectBuilder builder = Json.createObjectBuilder();
+            builder.add("headers",datasetAsJSONSchema);
```
these are schemaHeadersAsXML. please at least rename the variable as such. as for the further analysis of the code: why data exchange is done in xml? it would be much of a better separation of concern to leave such tricks to UI as it requires the data format change.

Laukkala replied:
Renamed variables to be more specific in commit [5f358af](https://github.com/teragrep/zep_01/pull/230/commits/5f358af6e947dff6c3afcf654996a1847f941313)

XML was used previously in draw() method to write headers directly to a paragraph, and having it be populated later by another update message. We removed draw() method to have the UI only receive data via one type of message, and headers are now being sent in JSON format

I don't think the XML formatted headers are used for anything anymore, so they were deleted in commit [2f56c1b](https://github.com/teragrep/zep_01/pull/230/commits/2f56c1bf36c49a97174b96122ec7eadded2a70c1)

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTTableDatasetNg.java:
```diff
@@ -249,40 +172,56 @@ public void setParagraphDataset(Dataset<Row> rowDataset) {
+
+        JsonObject headers;
+        // header formatting
+        if(datasetAsJSONFormattedSchema != ""){
```
why first convert from dataset.schema which is type of structype to json using iterator and builder, then to string then back to json using a reader?

Dataset has method schema() which returns StructType which inherits from DataType which has json() method that produces json with on hit wonder call 

https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/Dataset.html#schema()

https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/types/DataType.html#json()

Wouldn't this do?

Perhaps DTHeader should be gone by then.

Laukkala replied:
Schema.json() includes additional fields for each header, while we only need its name. Looks like the iterator was created to remove these fields first. Example of Schema.json() output for a schema with a single header:
`{"type":"struct","fields":[{"name":"_time","type":"timestamp","nullable":false,"metadata":{}}]}`

We can turn the JSONschema field from a String into a JsonArray to avoid double reading though. [a0a5dc7](https://github.com/teragrep/zep_01/pull/230/commits/a0a5dc7c32cce22173d059488a739096c5c45674)

kortemik commented on zeppelin-common/src/main/java/com/teragrep/zep_01/common/ValidatedMessage.java:
```diff
@@ -0,0 +1,87 @@
+
+package com.teragrep.zep_01.common;
+
+import com.google.gson.internal.LinkedTreeMap;
```
why gson and jakarta-json are both present?

Laukkala replied:
Leftover Exception declared in method signature. Removed as JsonExceptions are not thrown in current implementation. [ad8c309](https://github.com/teragrep/zep_01/pull/230/commits/ad8c30954cc399e94cb5a645e58d9adc73593719)

kortemik commented on zeppelin-common/src/main/java/com/teragrep/zep_01/common/ValidatedMessage.java:
```diff
@@ -0,0 +1,87 @@
+   * Checks that the message provided contains all of the required keys for a PARAGRAPH_UPDATE_RESULT message. Returns true if message is valid, false if not.
+   * @param messageToValidate Message object to be validated
+   */
+  public ValidatedMessage(Message messageToValidate){
```
this validates already the message which goes through the AjaxRequestWatcher, why double verification is done?

Laukkala replied:
Validation at the time the message is parsed is done so that we can respond with an error message if UI provides faulty parameters. 

I think AjaxRequestWatcher's validation would simply ignore invalid requests without any notification about faulty parameters, which might be confusing from the UI's perspective.

NotebookServer.onMessage() has a try-catch clause which is supposed to send an ERROR_INFO message to UI if an exception is encountered, but AjaxRequestWatcher's Exceptions are caught, logged and then discarded here: https://github.com/teragrep/zep_01/blob/1a0c747b26083de11d0064ff0f54a413740a1fbb/zeppelin-interpreter/src/main/java/com/teragrep/zep_01/display/AngularObject.java#L195 so Exceptions there cannot be sent to UI directly, at least without significant refactoring.

If we remove ValidatedMessage validation, UI will not be informed about bad requests, If we remove AjaxRequestWatchers validation, parsing a JsonObject can result in a NullPointerException, especially if we ever need to update AjaxRequest AngularObjects from another place apart from NotebookServer's Message objects. 

Should one or another of the validation steps be removed, and which one if so?

kortemik replied:
the way that the observer pattern is made here does not propagate the errors back like seen on the found snippet so it either requires refactoring of the observer pattern so that all exceptions get back here and error codes go back to the UI.

I would recommend that but to speed up the development a bit create an issue about that and validate it in the ValidatedMessage with the same logic that is in the watcher so we get the errors to the UI.

Laukkala replied:
Added issue #240 based on comments here.

In case of invalid data encountered in ValidatedMessage.isValid(), NotebookServer.updateParagraphResult() throws a BadRequestException, which should be caught by NotebookServer.onMessage() and result in an ERROR_INFO message being sent to UI.

kortemik commented on zeppelin-server/src/main/java/com/teragrep/zep_01/socket/NotebookServer.java:
```diff
@@ -1096,18 +1101,58 @@ public void onSuccess(Note newNote, ServiceContext context) throws IOException {
+      // Build an interpreterGroupId based on given user and note Id.
+      // InterpreterGroupId is used to find the correct AngularObjectRegistry instance containing the DTTableDatasetNG object we want to pass the search, length, start and draw values to.
+
+      final String interpreterGroupId = "spark-"+conn.getUser()+"-"+noteId;
```
i wonder if there is available such information ready made, building such is errorprone and does not work for others than spark

Laukkala replied:
InterpreterGroupId can be acquired from a given Paragraph: [e9a56f0](https://github.com/teragrep/zep_01/pull/230/commits/e9a56f07ca8f4038671f38c747f8a582a294b8da)
There are a lot of opportunities for zeppelin to return nulls while fetching the InterpreterGroupId. If an error occurs, a Bad Request response is sent to UI.

kortemik commented on zeppelin-server/src/main/java/com/teragrep/zep_01/socket/NotebookServer.java:
```diff
@@ -1096,18 +1101,58 @@ public void onSuccess(Note newNote, ServiceContext context) throws IOException {
+                                     Message fromMessage) throws IOException {
+    ValidatedMessage validatedMessage = new ValidatedMessage(fromMessage);
+    if(validatedMessage.isValid()){
+      // Casting is required to get Message parameters in correct format, as GSON parses all numbers as Doubles, and Message.get() returns a generic Object.
```
issue would be in place for removing such gson awkwardness

Laukkala replied:
Will require replacing Message class's usage of GSON with jakarta-json, as well as refactoring serializeMessage() and deserializeMessage() methods. Created new issue #239

kortemik reviewed (changes requested):
quite much, @eemhu please look into the PRs more in detail

Laukkala committed:
Changed Headers in DTHeader from JsonObject to JsonArray.

Laukkala committed:
Removed references to Gson and Gson.JsonSyntaxException from DTTableDatasetNg as Gson is no longer used.

Laukkala committed:
Removed unnecessary JsonException from ValidatedMessage.isValid signature, as the Exception type is not thrown anymore.

Laukkala committed:
Moved validation of AJAXRequest to AJAXRequestWatcher from DTTableDatasetNg.handeAJAXRequest()

Laukkala committed:
Renamed variables

Laukkala committed:
Removed schemaHeadersXml from DTTableDatasetNg as it had no usages apart from setting the variable in setParagraphDataset()

Laukkala committed:
Changed schemaHeadersJson from a String to a JsonArray to avoid double read of Json

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/AJAXRequestWatcher.java:
```diff
@@ -45,48 +45,50 @@
 
     @Override
-    public void watch(Object o, Object o1, InterpreterContext interpreterContext) {
+    public void watch(Object o, Object o1, InterpreterContext interpreterContext) throws JsonException {
```
I think such broken request should return 400 (bad request) to the UI it is too late here to throw exception on the watch. i wonder if watch can do other than try to understand the request and if it does not, it can log a technical error about that such malformed was encountered and not proceed any further. it is the responsibility of the notifier not the one being notified that the notifications are correct. this is a common observer pattern here. could we change it so that if such a malformed request ends up here it would just be logged as i think throwing on the watch won't return 400 for the UI?

Laukkala replied:
Added logging in commit [cb763c7](https://github.com/teragrep/zep_01/pull/230/commits/cb763c768bd7eea09ca5165263075e300b84abb6)
JsonException can occur when reading the request String into a JSON object using jakarta-json's Jsonreader. This should mean that the provided String is unparseable, so I added a catch clause for it too.

kortemik commented on spark/dpl-interpreter/src/main/java/com/teragrep/pth_07/ui/elements/table_dynamic/DTHeader.java:
```diff
@@ -61,10 +64,16 @@
+        this.schema = schema;
+    }
+
+    public String xml() throws TransformerException, ParserConfigurationException {
```
why xml() is still present even though it is no longer used?

Laukkala replied:
Removed in commit [202bb1f](https://github.com/teragrep/zep_01/pull/230/commits/202bb1f019e7c397fb6f8079863cdeb3cd0b6182)

Laukkala committed:
InterpreterGroupId is now generated from Paragraph information instead of being built based on message sender information

Laukkala committed:
Added error logging when either invalid JSON data or unparseable JSON data is received by AJAXRequestWatcher.

Laukkala committed:
Removed XML parsing from DTHeader as it had no usages.
